core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -407,30 +419,106 @@ public void index(Index index) {
         }
     }
 
+    private boolean canOptimizeAddDocument(Index index) {
+        if (index.getAutoGeneratedIdTimestamp() >= 0) {
+            switch (index.origin()) {
+                case PRIMARY:
+                    boolean value = (index.version() == Versions.MATCH_DELETED && index.versionType() == VersionType.INTERNAL)
+                        || (index.version() == Versions.MATCH_ANY && index.versionType() == VersionType.INTERNAL); // create request
+                    assert value; // we assert that since this was possible before and there might be some violations in the translog etc
```

Comment:
I think the comment needs some fixing - old translogs will never get here, as the `getAutoGeneratedIdTimestamp` will be -1.